nowcasting_dataset/data_sources/optical_flow/optical_flow_data_source.py

```diff
@@ -0,0 +1,267 @@
+            data=predictions,
+            dims={
+                "time_index": satellite_data.dims["time_index"],
+                "x_index": satellite_data.dims["x_index"],
+                "y_index": satellite_data.dims["y_index"],
+                "channels_index": satellite_data.dims["channels_index"],
+            },
+            coords={
+                "time_index": satellite_data.coords["time_index"],
+                "x_index": satellite_data.coords["x_index"],
+                "y_index": satellite_data.coords["y_index"],
+                "channels_index": satellite_data.coords["channels_index"],
+            },
+        )
+
+        return dataarray
+
+    def _get_previous_timesteps(
+        self,
+        satellite_data: xr.DataArray,
+        t0_dt: pd.Timestamp,
+    ) -> xr.DataArray:
+        """
+        Get timestamp of previous
+
+        Args:
+            satellite_data: Satellite data to use
+            t0_dt: Timestamp
+
+        Returns:
+            The previous timesteps
+        """
+        satellite_data = satellite_data.where(satellite_data.time <= t0_dt, drop=True)
+        return satellite_data
+
+    def _get_number_future_timesteps(
+        self, satellite_data: xr.DataArray, t0_dt: pd.Timestamp
+    ) -> int:
+        """
+        Get number of future timestamps
+
+        Args:
+            satellite_data: Satellite data to use
+            t0_dt: The timestamp of the t0 image
+
+        Returns:
+            The number of future timesteps
+        """
+        satellite_data = satellite_data.where(satellite_data.time > t0_dt, drop=True)
+        return len(satellite_data.coords["time_index"])
+
+    def _compute_and_return_optical_flow(
+        self,
+        satellite_data: xr.DataArray,
+        t0_dt: pd.Timestamp,
+    ) -> xr.DataArray:
+        """
+        Compute and return optical flow predictions for the example
+
+        Args:
+            satellite_data: Satellite DataArray
+            t0_dt: t0 timestamp
+
+        Returns:
+            The Tensor with the optical flow predictions for t0 to forecast horizon
+        """
+
+        # Get the previous timestamp
+        future_timesteps = self._get_number_future_timesteps(satellite_data, t0_dt)
+        satellite_data: xr.DataArray = self._get_previous_timesteps(
+            satellite_data,
+            t0_dt=t0_dt,
+        )
+        prediction_block = np.zeros(
+            (
+                future_timesteps,
+                self.image_size_pixels,
+                self.image_size_pixels,
+                satellite_data.sizes["channels_index"],
+            )
+        )
+        for prediction_timestep in range(future_timesteps):
+            for channel in range(0, len(satellite_data.coords["channels_index"]), 4):
+                # Optical Flow works with RGB images, so chunking channels for it to be faster
+                channel_images = satellite_data.sel(channels_index=slice(channel, channel + 3))
```

Comment:
The inner `for` loop is incrementing is steps of _4_ channels. But `channel_images` only includes 3 channels.  Does this mean that we're missing every 4th channel?!  Or have I misunderstood (entirely possible!)